Content.Server/GameObjects/EntitySystems/Atmos/GasTileOverlaySystem.cs

```diff
@@ -0,0 +1,350 @@
+            }
+        }
+
+        /// <summary>
+        ///     Checks whether the overlay-relevant data for a gas tile has been updated.
+        /// </summary>
+        /// <param name="gam"></param>
+        /// <param name="oldTile"></param>
+        /// <param name="indices"></param>
+        /// <param name="overlayData"></param>
+        /// <returns>true if updated</returns>
+        private bool TryRefreshTile(GridAtmosphereComponent gam, GasOverlayData oldTile, MapIndices indices, out GasOverlayData overlayData)
+        {
+            var tile = gam.GetTile(indices);
+            var tileData = new List<GasData>();
+
+            for (var i = 0; i < Atmospherics.TotalNumberOfGases; i++)
+            {
+                var gas = Atmospherics.GetGas(i);
+                var overlay = Atmospherics.GetOverlay(i);
+                if (overlay == null || tile.Air == null) continue;
+
+                var moles = tile.Air.Gases[i];
+
+                if (moles == 0.0f || moles < gas.GasMolesVisible) continue;
+
+                var data = new GasData(i, MathF.Max(MathF.Min(1, moles / gas.GasMolesVisibleMax), 0f));
+                tileData.Add(data);
+            }
+
+            overlayData = new GasOverlayData(tile.Hotspot.State, tile.Hotspot.Temperature, tileData.Count == 0 ? null : tileData.ToArray());
+
+            if (overlayData.Equals(oldTile))
+            {
+                return false;
+            }
+            
+            return true;
+        }
+
+        /// <summary>
+        ///     Get every chunk in range of our entity that exists, including on other grids.
+        /// </summary>
+        /// <param name="entity"></param>
+        /// <returns></returns>
+        private IEnumerable<GasOverlayChunk> GetChunksInRange(IEntity entity)
+        {
+            var entityTile = _mapManager
+                .GetGrid(entity.Transform.GridID)
+                .GetTileRef(entity.Transform.GridPosition)
+                .GridIndices;
+
+            // This is the max in any direction that we can get a chunk (e.g. max 2 chunks away of data).
+            var (maxXDiff, maxYDiff) = ((int) (UpdateRange / ChunkSize) + 1, (int) (UpdateRange / ChunkSize) + 1);
+            
+            var worldBounds = Box2.CenteredAround(entity.Transform.WorldPosition,
+                new Vector2(20.0f, 20.0f));
+
+            foreach (var grid in _mapManager.FindGridsIntersecting(entity.Transform.MapID, worldBounds))
+            {
+                if (!_overlay.TryGetValue(grid.Index, out var chunks))
+                {
+                    continue;
+                }
+                
+                for (var x = -maxXDiff; x <= maxXDiff; x++)
+                {
+                    for (var y = -maxYDiff; y <= maxYDiff; y++)
+                    {
+                        var chunkIndices = GetGasChunkIndices(new MapIndices(entityTile.X + x * ChunkSize, entityTile.Y + y * ChunkSize));
```

Comment:
Aren't these calculations completely wrong if the grid *isn't* the grid the player is currently on?

Reply:
... yes. AHHHHHHHHHHH.

Do you have any ideas that don't involve iterating over every single chunk on the grid?

Reply:
Convert the entity position to tile pos relative to the grid and then do the same math from there?

Reply:
Yeah I overthought that, new version looks fine when I aghost ingame and move off grid.